Adrian AI Materials (#5)
Ada-Developers-Academy/independent-ai-topics: working-with-openai/intro-to-openai.md

# Introduction to the OpenAI API

## Goals

> We explored how we can use ChatGPT to understand existing APIs and integrate them into our projects, but what if we wanted to leverage ChatGPT itself? In this lesson, we'll learn about the OpenAI API - the API behind ChatGPT itself - and how to set it up in a way that brings ChatGPT's many powerful tools right to our doorstep.

Reviewer: I believe we ended up cutting materials that explore existing APIs, so we could reword the first sentence to touch on using it for generating ideas, writing code, & debugging.

```suggestion
We explored how we can use tools built on OpenAI's offerings like ChatGPT and Copilot to generate ideas and to write and review code, but what if we wanted to leverage ChatGPT itself in our projects? In this lesson, we'll learn about the OpenAI API - the API behind ChatGPT itself - and how to set it up in a way that brings ChatGPT's many powerful tools right to our doorstep.
```

Our goals for this lesson are to:
- Understand what the OpenAI API is and how we can use it
- Understand how the OpenAI API is different than ChatGPT
- Set up an OpenAI account 
- Create an OpenAI environment on Postman that we can use for testing

> - Install the OpenAI Python library

Reviewer: Since we are covering just the first 2 points in this lesson, we may want to divide this up. Some options are framing the other points as a high level roadmap of what's coming next, or removing those bullets entirely

```suggestion
Our goals for this lesson are to:
- Understand what the OpenAI API is and how we can use it
- Understand how the OpenAI API is different than ChatGPT

To give a peek at our roadmap, the following lessons will cover topics like:
- Setting up an OpenAI account 
- Creating an OpenAI environment on Postman that we can use for testing
- Installing and using the OpenAI Python library
```

or

```suggestion
Our goals for this lesson are to:
- Understand what the OpenAI API is and how we can use it
- Understand how the OpenAI API is different than ChatGPT
```

## Vocabulary and Synonyms

| Vocab | Definition | Synonyms | How to Use in a Sentence |
| ----- | ---------- | -------- | ------------------------ |
| OpenAI | OpenAI is an organization that specializes in Artificial Intelligence research and deployment | | | 
| OpenAI API | An API created by OpenAI to allow users to access and deploy AI models within their own applications. | | |
| ChatGPT | A web app created by OpenAI that provides an interface for users to interact directly with AI models | | |

> | OpenAI Model | An AI program that has been trained on OpenAI data sets to achieve a specific purpose. | | |

Reviewer: I think we can take out the `Synonyms` column in this case since it isn't applicable for these terms. I'm also thinking we could remove the ChatGPT entry since students will have read up on it some and worked with it in previous lessons. We could still include it in one of the example sentences, maybe something like "ChatGPT is a tool that is built using the OpenAI APIs."?

```suggestion
## Vocabulary

| Vocab | Definition | How to Use in a Sentence |
| ----- | ---------- | ------------------------ |
| OpenAI | OpenAI is an organization that specializes in Artificial Intelligence research and deployment | |
| OpenAI API | An API created by OpenAI to allow users to access and deploy AI models within their own applications. | |
| OpenAI Model | An AI program that has been trained on OpenAI data sets to achieve a specific purpose. | |
```

## What is the OpenAI API?

Just as an API allows us to interact with and access the functions, data and other features of a particular service, the OpenAI API allows us to access the features provided by ChatGPT's parent organization, OpenAI. As a result, the OpenAI API can be useful in integrating a variey of AI tools into the applications we create. 

Some practical applications include but are not limited to:
- Creating a chatbot or virtual assistant

> - Content generation

Reviewer: It might be helpful to include a really short example here since "content generation" could include a lot of things.

- Image recognition
- Language translation
- Generating recommendations based off user input
- Providing accessibility considerations such as text to speech or speech to text

## The OpenAI API vs ChatGPT

In order for us to work all of these tools into our projects, it is important to understand the distinction between ChatGPT and the OpenAI API itself. As we have discussed before, ChatGPT is an online platform that is open for use to the general public. It is a (mostly) free service and it is primarily used for generating text, retrieving information and engaging in dialogue with the LLM on which it has been trained.

On the other hand, the OpenAI API is an API that allows developers to integrate GPT models into their applications. We will discuss GPT models in a bit, but briefly, they are LLMs that have been trained for specific actions, whether that be retrieving information or generating content, etc. The OpenAI API allows us to access these different models and use them in our development.

> If we look at ChatGPT itself, we see that the free version is built using the GPT 3.5 model from the OpenAI API. It is built specifically to answer questions or provide recommendations/generate content based off user text input. The paid version is built similarly but uses the better trained GPT 4.0 model for its conversational piece and also incorporates other models that can handle input in the form of images, video or audio. If we wanted to, we could use the OpenAI API to build our own pseudo ChatGPT, but since that has already been created, it is up to us to find other exciting ways to leverage OpenAI's API within our own projects.

Reviewer: Small adjustment to list.

```suggestion
If we look at ChatGPT itself, we see that the free version is built using the GPT 3.5 model from the OpenAI API. It is built specifically to answer questions, provide recommendations, or generate content based off user text input. The paid version is built similarly but uses the better trained GPT 4.0 model for its conversational piece and also incorporates other models that can handle input in the form of images, video or audio. If we wanted to, we could use the OpenAI API to build our own pseudo ChatGPT, but since that has already been created, it is up to us to find other exciting ways to leverage OpenAI's API within our own projects.
```

## Setting up the OpenAI API

### !callout-warn

## The OpenAI is Not Free at this Time
Unfortunately, the OpenAI is not currently a free API. It is priced according to which model you are using and how often you are using it. New users do currently get $5 worth of credits to begin experimenting which is tied to the phone number you provide when signing up. 

> ### !end-callout

Reviewer: Since this "Setting up the OpenAI API" material is on its own page now, how do you feel about replacing this section with a Summary heading that contains just a few sentences gathering the top couple things you'd want a student to take away from this page?

Author: Done!
